recipes/entt/3.x.x/conanfile.py

```diff
@@ -19,6 +19,10 @@ def _source_subfolder(self):
         return "source_subfolder"
 
     def configure(self):
+        # FIXME: Here we are implementing a workaround because ConanCenter is not generating any configuration with
+        #   C++17 support (either supported by default by a compiler or using 'compiler.cppstd=17' in the Conan profile)
+        #   and ConanCenter requires that at least one package is generated. Once ConanCenter uses a profile
+        #   with C++17 support, only a raw call to `tools.check_mis_cppstd(self, "17")` will be required.
```

Comment:
Why this comment by the way? All CCI recipes requiring at least C++14 have this kind of pattern.

Reply:
The C++ standard should be known by Conan in order to select the proper binary. Once we start to generate here binaries for stable C++ standards, recipes might look like the following:

 * A recipe that needs C++17 to build, but the API is C++11 compatible:

```python
class Recipe:

def validate(self):
    tools.check_min_cppstd(self, "17")  # This is needed to build

def package_id(self):
    # If the API is compatible for >=C++11, we can fallback to any binary available
    for cppstd in ("11", "14", "17", "20"):
        compatible_pkg = self.info.clone()
        compatible_pkg.settings.compiler.cppstd = cppstd
        self.compatible_packages.append(compatible_pkg)
```

So, a profile without `compiler.cppstd` will be suitable to consume this library if the default for the compiler is at least C++11, but only a profile with explicit `compiler.cppstd=17` or `compiler.cppstd=20` will be able to generate this library. 

Behind the scenes, Conan keeps the exception raises by the `check_min_cppstd` and only raise it if the command is trying to build the library, otherwise this package remains in the graph with `package_id = INVALID` (that would be the output for `conan info` command), and it will fall back to the compatible package if we need a binary to install.

Reply:
With this implementation, you can mix binaries from different C++ standards because they are declared **compatible** not just because a _hidden_ `CXX_STANDARD 17` in the `CMakeLists.txt` was able to generate the binary. I think it is preferred to make it explicit.

Reply:
And if default standard of a compiler version is not the minimum standard required by the library, but still supports enough features of this standard for this library?

Reply:
> And if default standard of a compiler version is lower than the minimum standard required by the library, but still supports enough features of this standard for this library? Will it work if standard is not explicitly set in profile?
> 

We cannot handle this situation, if the recipe says `tools.check_min_cppstd(self, "17")` it will fail if the compiler doesn`t implement that standard (if it is declared stable). If it only needs some features that are available in previous versions, the recipe will need some handcrafted checks... but I would try to avoid them as much as possible (in my 🦄  world).

> Moreover, AFAIK `tools.check_min_cppstd(self, "17")` raises if standard is not set in profile, so it requires modifications in this function.
> 

No, it doesn't (https://github.com/conan-io/conan/blob/develop/conans/client/tools/settings.py#L5). You saying that is a symptom of the workaround we are using in the recipe to bypass the fact that ConanCenter is not trying any profile with C++14/17/20 support.

> What happen if a consumer tries to build from source with an unsupported compiler? Currently, it gracefully fails, while with this implementation it will fail at some point during build()

It will fail gracefully, have a look at the docs about `validate()` function: https://docs.conan.io/en/latest/reference/conanfile/methods.html?highlight=validate#validate

Reply:
If cppstd is not set, `tools.check_min_cppstd` raises if default standard as estimated by conan is not sufficient. Currently, `tools.check_min_cppstd(self, 17)` fails with all versions of Visual Studio.

Reply:
If `compiler.cppstd` is not set in the profile it will fail **to build**, yes. And it is right because the recipe says C++17 is required and those compilers don't enable C++17 support by default. Activating the standard in the `CMakeLists.txt` file is something to avoid, it is hiding changes in the ABI from the package manager.

But, by adding the `compatible_packages` feature, Conan will be able to retrieve a compatible binary that was compiled with an explicit `compiler.cppstd` value when using a profile without it.


> I believe that, if your first answer in the previous message was implemented as is in CCI recipes, it would break a lot of them (a lot of C++11 recipes work with gcc 4.9/5, while C++11 is not the default standard of those versions).

As they are now, those recipes are wrong (because of c3i profiles), they are activating C++11 inside the `CMakeLists.txt`, and some of the header-only ones are activating it inside the `test_package/CMakeLists.txt` (if you add them to your graph, Conan retrieves a package and then your consumer fails to compile without further notice). If you modify those recipes as the draft in the first commit they will fail to build if you do not provide a compatible configuration in the profile (nothing hidden) but you still can consume them with a lower C++ standard if the API is compatible and the binary is available (it was compiled with a higher C++ standard).

Maybe this requires a written example

Reply:
It's not wrong to activate C++11 in CMakeLists.txt, it's what all C++11 libraries in the world do. I don't understand your point. It fails in test_package if C++ standard is not set, because conan doesn't provide a way (or if it does, nobody use it in CCI recipes) to easily propage this information in generators.

Reply:
I've written an example with a (opinionated) dramatization here: https://github.com/conan-io/examples/pull/73. I know _every_ library activates the C++ standard in the `CMakeLists.txt` but IMHO it is not the best place to do it. We can move this conversation there (or to another repository) with an example, it is easier for me to show what I want to show.

I will remove the comment in this recipe as this is something we cannot fix right now, but I really think user experience and ConanCenter would benefit from that approach. Please have a look at the recipes in that PR. Thanks!

Reply:
Ok thanks, I'll take a look ;)